File: problems/2026/04/2026-04-15-leetcode-102-二叉树的层序遍历/SolutionLC102.py
from __future__ import annotations

from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    """LeetCode 102：二叉树的层序遍历（BFS）"""

    def level_order(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        res: List[List[int]] = []
        q: deque[TreeNode] = deque([root])
        while q:
            sz = len(q)
            level: List[int] = []
            for _ in range(sz):
                node = q.popleft()
                level.append(node.val)
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
            res.append(level)
        return res


if __name__ == "__main__":
    sol = Solution()
    tree = TreeNode(
        3,
        TreeNode(9),
        TreeNode(20, TreeNode(15), TreeNode(7)),
    )
    assert sol.level_order(tree) == [[3], [9, 20], [15, 7]]
    assert sol.level_order(TreeNode(1)) == [[1]]
    assert sol.level_order(None) == []
    print("示例均通过")

File: problems/2026/04/2026-04-15-leetcode-102-二叉树的层序遍历/SolutionLC102.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * LeetCode 102：二叉树的层序遍历（BFS 按层处理）
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class SolutionLC102 {

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Deque<TreeNode> q = new ArrayDeque<>();
        q.offer(root);
        while (!q.isEmpty()) {
            int sz = q.size();
            List<Integer> level = new ArrayList<>(sz);
            for (int i = 0; i < sz; i++) {
                TreeNode node = q.poll();
                level.add(node.val);
                if (node.left != null) {
                    q.offer(node.left);
                }
                if (node.right != null) {
                    q.offer(node.right);
                }
            }
            res.add(level);
        }
        return res;
    }

    public static void main(String[] args) {
        SolutionLC102 solution = new SolutionLC102();
        // [3,9,20,null,null,15,7]
        TreeNode root = new TreeNode(3,
                new TreeNode(9),
                new TreeNode(20, new TreeNode(15), new TreeNode(7)));
        System.out.println(solution.levelOrder(root));
    }
}
